## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-zoned.sh | :x: [more...](#0fd6cd80-1f65-4378-9264-f88af91a1e5b) |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#98fb094f-1782-4bac-a641-f8073ed675a2) |
<a name="98fb094f-1782-4bac-a641-f8073ed675a2"></a>
### 12233-51d078a/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:31196
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp b9dbc3a74ce1:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 42 seconds
(4/15) Running aws-sdk-ruby tests ... done in 4 seconds
(5/15) Running awscli tests ... FAILED in 32 seconds
{
  "name": "awscli",
  "duration": 2737,
  "function": "aws --endpoint-url http://minio-c3.minio.io:31196 s3api copy-object --bucket awscli-mint-test-bucket-7229 --key datafile-1-kB-copy --copy-source awscli-mint-test-bucket-7229/datafile-1-kB\n",
  "status": "FAIL",
  "error": "Hash mismatch expected 084e1383b70fb0c51acc680fef370023, got ac57de7156d7fc25ac1a65f81fa3989b"
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 54 seconds
(7/15) Running minio-dotnet tests ... done in 45 seconds
(8/15) Running minio-go tests ... FAILED in 2 minutes and 41 seconds
{
  "args": {},
  "duration": 676,
  "error": "At least one of the pre-conditions you specified did not hold",
  "function": "CopyObjectPart(destination, source)",
  "message": "CopyObjectPart call failed",
  "name": "minio-go: testUnencryptedToSSES3CopyObjectPart",
  "status": "FAIL"
}
(8/15) Running minio-java tests ... FAILED in 2 minutes and 1 seconds
{
  "name": "minio-java",
  "function": "copyObject()",
  "args": "[match etag]",
  "duration": 134,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = PreconditionFailed, message = At least one of the pre-conditions you specified did not hold, bucketName = minio-java-test-2rn3ing, objectName = minio-java-test-o4hcvf-copy, resource = /minio-java-test-2rn3ing/minio-java-test-o4hcvf-copy, requestId = 167C432A2224AC45, hostId = 47b5c813-6ce9-4a21-bd6e-623f136b099f)\nrequest={method=PUT, url=http://minio-c3.minio.io:31196/minio-java-test-2rn3ing/minio-java-test-o4hcvf-copy, headers=x-amz-copy-source-if-match: 71cff0a060f852067e443ad1e24ae26c-1\nx-amz-copy-source: /minio-java-test-keh5if/minio-java-test-o4hcvf\nHost: minio-c3.minio.io:31196\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210505T194334Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210505/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-copy-source;x-amz-copy-source-if-match;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=412, headers=Accept-Ranges: bytes\nContent-Length: 416\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nETag: \"71cff0a060f852067e443ad1e24ae26c\"\nLast-Modified: Wed, 05 May 2021 19:43:34 GMT\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 167C432A2224AC45\nX-Xss-Protection: 1; mode=block\nDate: Wed, 05 May 2021 19:43:34 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.execute(MinioClient.java:563), io.minio.MinioClient.executePut(MinioClient.java:904), io.minio.MinioClient.copyObject(MinioClient.java:1232), FunctionalTest.testCopyObjectMatchETag(FunctionalTest.java:1850), FunctionalTest.copyObject(FunctionalTest.java:2016), FunctionalTest.runObjectTests(FunctionalTest.java:3757), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(8/15) Running minio-js tests ... done in 47 seconds
(9/15) Running minio-py tests ... done in 2 minutes and 48 seconds
(10/15) Running s3cmd tests ... FAILED in 5 seconds
{
  "name": "s3cmd",
  "duration": "2637",
  "function": "test_put_object_multipart",
  "status": "FAIL",
  "error": "WARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Too many failures. Giving up on '/mint/data/datafile-65-MB'\nERROR: \nUpload of '/mint/data/datafile-65-MB' part 1 failed. Use\n  /usr/local/bin/s3cmd abortmp s3://s3cmd-test-bucket-7656/s3cmd-test-object-23548 ac6cf836-2dec-4d20-95d5-9c750378f560\nto abort the upload, or\n  /usr/local/bin/s3cmd --upload-id ac6cf836-2dec-4d20-95d5-9c750378f560 put ...\nto continue the upload.\nERROR: Upload of '/mint/data/datafile-65-MB' failed too many times (Last reason: )"
}
(10/15) Running s3select tests ... done in 6 seconds
(11/15) Running security tests ... done in 0 seconds

Executed 11 out of 15 tests successfully.
```
<a name="0fd6cd80-1f65-4378-9264-f88af91a1e5b"></a>
### 12233-51d078a/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:32425
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 5995ce52e6a6:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 3 seconds
(5/15) Running awscli tests ... FAILED in 39 seconds
{
  "name": "awscli",
  "duration": 3302,
  "function": "aws --endpoint-url http://minio-dev4.minio.io:32425 s3api head-object --bucket awscli-mint-test-bucket-6487 --key datafile-1-kB\n",
  "status": "FAIL",
  "error": "StorageClass was not applied"
}
(5/15) Running healthcheck tests ... done in 1 seconds
(6/15) Running mc tests ... done in 1 minutes and 1 seconds
(7/15) Running minio-dotnet tests ... done in 44 seconds
(8/15) Running minio-go tests ... FAILED in 1 minutes and 53 seconds
{
  "args": {
    "bucketName": "minio-go-test-02fm0ejey2xo0ytj",
    "objectName": "i5zse55rc6qaqccmghqiq4q3dmzown",
    "prefix": "",
    "recursive": ""
  },
  "duration": 546,
  "function": "ListObjectVersions(bucketName, prefix, recursive)",
  "message": "Unexpected number of DeleteMarker elements in listing objects",
  "name": "minio-go: testListObjectVersions",
  "status": "FAIL"
}
(8/15) Running minio-java tests ... FAILED in 27 seconds
{
  "name": "minio-java",
  "function": "listObjects()",
  "args": "[bucket, versions]",
  "duration": 1949,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = BucketNotEmpty, message = The bucket you tried to delete is not empty. You must delete all versions in the bucket., bucketName = minio-java-test-2nqj3pc, objectName = null, resource = /minio-java-test-2nqj3pc, requestId = 167C430C07F75C10, hostId = 01609cea-33ab-4ce5-be05-1f5b90ddacf0)\nrequest={method=DELETE, url=http://minio-dev4.minio.io:32425/minio-java-test-2nqj3pc, headers=Host: minio-dev4.minio.io:32425\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210505T194124Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210505/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=409, headers=Accept-Ranges: bytes\nContent-Length: 373\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 167C430C07F75C10\nX-Xss-Protection: 1; mode=block\nDate: Wed, 05 May 2021 19:41:24 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.execute(MinioClient.java:563), io.minio.MinioClient.executeDelete(MinioClient.java:879), io.minio.MinioClient.removeBucket(MinioClient.java:2669), FunctionalTest.testListObjects(FunctionalTest.java:1416), FunctionalTest.listObjects(FunctionalTest.java:1442), FunctionalTest.runObjectTests(FunctionalTest.java:3747), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(8/15) Running minio-js tests ... FAILED in 27 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 7,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:79:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (internal/streams/readable.js:684:10) at endReadableNT (internal/streams/readable.js:1327:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(8/15) Running minio-py tests ... FAILED in 6 seconds
{
  "name": "minio-py:test_copy_object_etag_match",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-d68387bf-d114-4c48-b1ed-1f6f4f8b295d",
    "object_source": "b29df158-b21f-4ffd-8528-9b514fe4a5d5-source",
    "object_name": "b29df158-b21f-4ffd-8528-9b514fe4a5d5-copy",
    "conditions": {
      "set_match_etag": "43b5c38e7d6bbb42d05ee0c447181bee"
    }
  },
  "message": "S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-d68387bf-d114-4c48-b1ed-1f6f4f8b295d, request_id: 167C4313B97BC35D, host_id: 01609cea-33ab-4ce5-be05-1f5b90ddacf0, bucket_name: minio-py-test-d68387bf-d114-4c48-b1ed-1f6f4f8b295d",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 536, in test_copy_object_etag_match\n    _CLIENT.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 676, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-d68387bf-d114-4c48-b1ed-1f6f4f8b295d, request_id: 167C4313B97BC35D, host_id: 01609cea-33ab-4ce5-be05-1f5b90ddacf0, bucket_name: minio-py-test-d68387bf-d114-4c48-b1ed-1f6f4f8b295d\n",
  "duration": 697
}
(8/15) Running s3cmd tests ... done in 23 seconds
(9/15) Running s3select tests ... done in 10 seconds
(10/15) Running security tests ... done in 0 seconds

Executed 10 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/api-errors.go
+++ b/cmd/api-errors.go
@@ -1082,9 +1082,9 @@ var errorCodes = errorCodeMap{
 		HTTPStatusCode: http.StatusBadRequest,
 	},
 	ErrKMSNotConfigured: {
-		Code:           "InvalidArgument",
+		Code:           "NotImplemented",
 		Description:    "Server side encryption specified but KMS is not configured",
-		HTTPStatusCode: http.StatusBadRequest,
+		HTTPStatusCode: http.StatusNotImplemented,
 	},
 	ErrNoAccessKey: {
 		Code:           "AccessDenied",

--- a/cmd/erasure-server-pool.go
+++ b/cmd/erasure-server-pool.go
@@ -269,47 +269,6 @@ func (z *erasureServerPools) getPoolIdxExisting(ctx context.Context, bucket, obj
 	wg.Wait()
 
 	for i, err := range errs {
-		if err == nil {
-			return i, nil
-		}
-		if isErrObjectNotFound(err) {
-			// No object exists or its a delete marker,
-			// check objInfo to confirm.
-			if objInfos[i].DeleteMarker && objInfos[i].Name != "" {
-				return i, nil
-			}
-
-			// objInfo is not valid, truly the object doesn't
-			// exist proceed to next pool.
-			continue
-		}
-		return -1, err
-	}
-
-	return -1, toObjectErr(errFileNotFound, bucket, object)
-}
-
-// getPoolIdx returns the found previous object and its corresponding pool idx,
-// if none are found falls back to most available space pool.
-func (z *erasureServerPools) getPoolIdx(ctx context.Context, bucket, object string, size int64) (idx int, err error) {
-	if z.SinglePool() {
-		return 0, nil
-	}
-
-	errs := make([]error, len(z.serverPools))
-	objInfos := make([]ObjectInfo, len(z.serverPools))
-
-	var wg sync.WaitGroup
-	for i, pool := range z.serverPools {
-		wg.Add(1)
-		go func(i int, pool *erasureSets) {
-			defer wg.Done()
-			objInfos[i], errs[i] = pool.GetObjectInfo(ctx, bucket, object, ObjectOptions{})
-		}(i, pool)
-	}
-	wg.Wait()
-
-	for i, err := range errs {
 		if err != nil && !isErrObjectNotFound(err) {
 			return -1, err
 		}
@@ -319,19 +278,33 @@ func (z *erasureServerPools) getPoolIdx(ctx context.Context, bucket, object stri
 			if objInfos[i].DeleteMarker && objInfos[i].Name != "" {
 				return i, nil
 			}
+
 			// objInfo is not valid, truly the object doesn't
 			// exist proceed to next pool.
 			continue
 		}
-		// object exists at this pool.
 		return i, nil
 	}
 
-	// We multiply the size by 2 to account for erasure coding.
-	idx = z.getAvailablePoolIdx(ctx, size*2)
-	if idx < 0 {
-		return -1, toObjectErr(errDiskFull)
+	return -1, toObjectErr(errFileNotFound, bucket, object)
+}
+
+// getPoolIdx returns the found previous object and its corresponding pool idx,
+// if none are found falls back to most available space pool.
+func (z *erasureServerPools) getPoolIdx(ctx context.Context, bucket, object string, size int64) (idx int, err error) {
+	idx, err = z.getPoolIdxExisting(ctx, bucket, object)
+	if err != nil && !isErrObjectNotFound(err) {
+		return idx, err
 	}
+
+	if isErrObjectNotFound(err) {
+		// We multiply the size by 2 to account for erasure coding.
+		idx = z.getAvailablePoolIdx(ctx, size*2)
+		if idx < 0 {
+			return -1, toObjectErr(errDiskFull)
+		}
+	}
+
 	return idx, nil
 }
 
@@ -1039,6 +1012,14 @@ func (z *erasureServerPools) NewMultipartUpload(ctx context.Context, bucket, obj
 		return z.serverPools[0].NewMultipartUpload(ctx, bucket, object, opts)
 	}
 
+	ns := z.NewNSLock(minioMetaMultipartBucket, pathJoin(bucket, object, "newMultipartObject.lck"))
+	lkctx, err := ns.GetLock(ctx, globalOperationTimeout)
+	if err != nil {
+		return "", err
+	}
+	ctx = lkctx.Context()
+	defer ns.Unlock(lkctx.Cancel)
+
 	for idx, pool := range z.serverPools {
 		result, err := pool.ListMultipartUploads(ctx, bucket, object, "", "", "", maxUploadsList)
 		if err != nil {
@@ -1052,10 +1033,9 @@ func (z *erasureServerPools) NewMultipartUpload(ctx context.Context, bucket, obj
 		}
 	}
 
-	// We multiply the size by 2 to account for erasure coding.
-	idx := z.getAvailablePoolIdx(ctx, (1<<30)*2)
-	if idx < 0 {
-		return "", toObjectErr(errDiskFull)
+	idx, err := z.getPoolIdx(ctx, bucket, object, 1<<30)
+	if err != nil {
+		return "", err
 	}
 
 	return z.serverPools[idx].NewMultipartUpload(ctx, bucket, object, opts)

